Add a HTML/Javascript front-end example

66OJ66 opened the pull request:
Hi, as requested on #252, this PR adds a basic HTML/Javascript front-end

- [x] cargo test has been run and passes
- [x] documentation has been updated with relevant examples (if relevant)

Rather than just adding the `.html` and `.js` files to a new folder under _webauthn-rs/tutorial/frontend/javascript_, I've added 2 features (`wasm` and `javascript`) to the Axum example so it's easier to use either front-end. `wasm` is marked as a default feature. These features are mutually exclusive.

axum-extra is added as a dependency so the [SpaRouter](https://docs.rs/axum-extra/latest/axum_extra/routing/struct.SpaRouter.html) can be used. This also allowed me to simplify some of the code (e.g. remove `file.rs`).

The index_view() HTML has been moved into assets/wasm_index.html and I changed `/pkg/wasm_bg.wasm` to `/assets/wasm_bg.wasm` etc.
The rationale for this was I felt pkg doesn't really apply to HTML/Javascript, and I wanted to keep the directory names consistent. Feel free to revert this though.

Firstyear commented:
I'm happy with the rename of pkg to assets, but how is spaRouter actually accessing and serving the .wasm files? I can't see how it links to the built outputs. Am I missing something?

micolous commented on tutorial/server/axum/assets/auth.js:
```diff
@@ -0,0 +1,93 @@
+        return;
+    }
+
+    fetch('http://localhost:8080/register_start/' + username, {
```
`username` [needs to be escaped](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) here, and also below.

(it also looks like this is a bug with the Rust/WASM bindings, but that doesn't need to be addressed in this PR)

micolous commented on tutorial/server/axum/assets/js_index.html:
```diff
@@ -0,0 +1,20 @@
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>WebAuthn-rs Tutorial</title>
+	<script src="https://cdn.jsdelivr.net/npm/js-base64@3.7.4/base64.min.js"></script>
```
This file looks like it has a mix of tabs and spaces, so indentation is inconsistent

micolous replied:
As this pins a specific version of the package, can this use [subresource integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) and `crossorigin="anonymous"`

micolous commented on tutorial/server/axum/assets/auth.js:
```diff
@@ -0,0 +1,93 @@
+        })
+        .then((response) => {
+            if (response.ok){
+                console.log("Registered!");
```
Can success / failure be shown on the page as well?

ie: `document.getElementById('status').innerText = "Registered!";`

micolous commented on tutorial/server/axum/src/main.rs:
```diff
@@ -26,39 +20,15 @@ extern crate tracing;
 mod auth;
 mod startup;
 
+#[cfg(all(feature = "javascript", feature = "wasm"))]
```
What happens when neither feature is selected?

66OJ66 replied:
In that case, nothing would be served when navigating to **localhost:8080/** - that's why I marked the `wasm` feature as default

micolous commented on tutorial/server/axum/README.md:
````diff
@@ -0,0 +1,10 @@
+By default, it serves the WASM front-end ([located here](https://github.com/kanidm/webauthn-rs/tree/master/tutorial/wasm "located here")). You will need to build this yourself first then copy the output into the assets directory.
+
+If you want to use the HTML/Javascript front-end instead, run the following command
+```rust
````
This is `sh`, not `rust`.

66OJ66 replied:
Oops - corrected now

66OJ66 commented:
@Firstyear So the wasm_index.html file contains this:
```html
<script type="module">
        import init, { run_app } from './assets/wasm.js';
        async function main() {
           await init('./assets/wasm_bg.wasm');
           run_app();
        }
        main()
</script>
```

With the server running, when you navigate to **localhost:8080/** it:
1. Imports functions from /assets/wasm.js
2. Runs the main() function which retrieves the .wasm file from /assets


So you see this
![WASM Loading](https://user-images.githubusercontent.com/59021291/212054454-20ade4e5-eb54-4f56-877b-9de507ef0458.png)

then this
![WASM Loaded](https://user-images.githubusercontent.com/59021291/212054547-ca16d59a-3760-4cd6-8b82-1b1d3c8d2208.png)

Firstyear commented:
> @Firstyear So the wasm_index.html file contains this:

I meant more that I can't see what serves the *files* in the code here. You've created an "/assets" folder but I can't see it. Additionally, you have tutorial/server/axum/assets/wasm_index.html but wasm_bg.wasm isn't in that folder either? How does the axum server get to and serve the wasm_bg.wasm file itself?

66OJ66 commented:
> How does the axum server get to and serve the wasm_bg.wasm file itself?

Ah sorry - initially the WASM files had to be copied across manually into _tutorial/server/axum/assets_.

I've just added a new shell script to build and copy the WASM files across automatically (it's essentially the same as the existing `build.sh` script - I've just changed the input and output directories).

Firstyear commented:
@66OJ66 Thank you!
